Update smartplug.py to support dimming in HS220

JsChiSurf opened the pull request:
Switch functions essentially as a "plug" with the addition of supporting dimming,  which can be test for by verifying existence of the 'brightness' array value.

JsChiSurf committed:
Update smartplug.py to support dimming in HS220

Switch functions essentially as a "plug" with the addition to support for dimming, for which can be test for by verifying existence of 'brightness' array value.

JsChiSurf committed:
Attempt at updates to pass validator

JsChiSurf committed:
Maybe this time?  :-)

rytilahti commented:
Without having looked into this that much, I think that the dimmer is more like a bulb than it is a switch, and should therefore belong to smartbulb. Would that make any sense?

JsChiSurf commented:
It's tough since it's a hybrid. But, it seems to share more in common with the plug library than the bulb, with the only difference being the brightness setting. I have multiple HS200s (non-dimmable switch), and I wouldn't think to use the bulb library for them.

rytilahti commented:
Ok, sounds fair enough! I'm wondering how shall we do autodetection for these devices (maybe just checking if the brightness is exposed?). I'll add a couple of comments to the code now already though.

rytilahti commented on pyHS100/smartplug.py:
```diff
@@ -80,6 +80,54 @@ def state(self, value: str):
+    @property
+    def brightness(self) -> Optional[int]:
+        """
+        Current brightness of the device, if supported
```
Please add the range of returned values here.

rytilahti commented on pyHS100/smartplug.py:
```diff
@@ -80,6 +80,54 @@ def state(self, value: str):
+    def brightness(self, value: int):
+        """
+        Set the new switch brightness level
+
```
Maybe it is worth mentioning here, that this will turn the light on if it isn't already on?

rytilahti commented on pyHS100/smartplug.py:
```diff
@@ -80,6 +80,54 @@ def state(self, value: str):
+
+        if not isinstance(value, int):
+            raise ValueError("Brightness must be integer, not of %s.", type(value))
+        elif value > 0 and value <= 100:
```
Just to be sure, the values are between 1-100?

JsChiSurf replied:
Yes, this is correct, I specified valid values to be between 1 and 100.  I did not feel comfortable "assuming" that setting brightness to 0 would turn the light off, versus an explicit call by sending to the "off" command.  I can test this tonight by sending a value of 0 for brightness to see if the "relay_state" automatically switches to 0 when / if brightness is set to 0 before, and if so, allow 0 in the brightness range before this is merged.

I will also update the auto discovery & README.md files accordingly.

JsChiSurf replied:
Okay, I VPN'd home from work and can indeed confirm that allowing a brightness of 0 will NOT turn the switch off and in addition will NOT change whatever the current brightness level is (i.e. if brightness is set to 60, and a value of 0 is passed in, it will stay at 60).  Furthermore, the way things are currently coded, by allowing 0 as a brightness value, this would turn the light on to whatever its previous brightness value is set to, since 0 has no impact.  So, I think things are good they way they are.

rytilahti reviewed (commented):
Looks generally good to me, needs updating the auto discovery & README.md to include the new type, but otherwise this can be merged in my opinion.

JsChiSurf committed:
Add more detail to comment blocks

Make clear in requests for current brightness level the expected return values, and note that light will turn on when setting a brightness level, if not already on.  This makes clear that a state change request (turn_on) does NOT have to be made first when setting brightness.

JsChiSurf committed:
Update smartplug.py

JsChiSurf committed:
Update smartplug.py

rytilahti commented:
Looks good, let's get this merged. Thanks a lot!
